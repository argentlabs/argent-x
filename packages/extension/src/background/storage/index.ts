export * from "./interface"
export * from "./default"
