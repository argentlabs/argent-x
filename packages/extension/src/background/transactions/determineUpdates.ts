import type { Transaction } from "../../shared/transactions"
import { compareTransactions } from "../../shared/transactions"
import { getTransactionStatus } from "../../shared/transactions/utils"

export function getTransactionsStatusUpdate(
  oldTransactions: Transaction[],
  newTransactions: Transaction[],
): Transaction[] {
  return newTransactions.filter((newTransaction) => {
    const oldTransaction = oldTransactions.find((oldTransaction) =>
      compareTransactions(oldTransaction, newTransaction),
    )

    const {
      finality_status: oldFinalityStatus,
      execution_status: oldExecutionStatus,
    } = getTransactionStatus(oldTransaction)
    const {
      finality_status: newFinalityStatus,
      execution_status: newExecutionStatus,
    } = getTransactionStatus(newTransaction)

    return (
      !oldTransaction ||
      oldFinalityStatus !== newFinalityStatus ||
      oldExecutionStatus !== newExecutionStatus
    )
  })
}
