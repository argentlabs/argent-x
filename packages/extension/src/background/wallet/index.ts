import { Account, InvocationsDetails } from "starknet"

import { Address } from "@argent/x-shared"
import { ProgressCallback } from "ethers"
import { WalletAccountSharedService } from "../../shared/account/service/accountSharedService/WalletAccountSharedService"
import { PendingMultisig } from "../../shared/multisig/types"
import { Network } from "../../shared/network"
import { BaseSignerInterface } from "../../shared/signer/BaseSignerInterface"
import {
  ArgentAccountType,
  BaseMultisigWalletAccount,
  BaseWalletAccount,
  CreateAccountType,
  ImportedLedgerAccount,
  MultisigData,
  NetworkOnlyPlaceholderAccount,
  SignerType,
  WalletAccount,
} from "../../shared/wallet.model"
import { WalletAccountStarknetService } from "./account/WalletAccountStarknetService"
import { WalletBackupService } from "./backup/WalletBackupService"
import { WalletCryptoSharedService } from "./crypto/WalletCryptoSharedService"
import { WalletCryptoStarknetService } from "./crypto/WalletCryptoStarknetService"
import { WalletDeploymentStarknetService } from "./deployment/WalletDeploymentStarknetService"
import { WalletRecoverySharedService } from "./recovery/WalletRecoverySharedService"
import { WalletRecoveryStarknetService } from "./recovery/WalletRecoveryStarknetService"
import { WalletSessionService } from "./session/WalletSessionService"

export class Wallet {
  constructor(
    private readonly walletAccountSharedService: WalletAccountSharedService,
    private readonly walletAccountStarknetService: WalletAccountStarknetService,
    private readonly walletBackupService: WalletBackupService,
    private readonly walletCryptoSharedService: WalletCryptoSharedService,
    private readonly walletCryptoStarknetService: WalletCryptoStarknetService,
    private readonly walletDeploymentStarknetService: WalletDeploymentStarknetService,
    private readonly walletRecoverySharedService: WalletRecoverySharedService,
    private readonly walletRecoveryStarknetService: WalletRecoveryStarknetService,
    private readonly walletSessionService: WalletSessionService,
  ) {}

  // WalletAccountSharedService
  public async getDefaultAccountName(
    networkId: string,
    type: CreateAccountType,
  ) {
    return this.walletAccountSharedService.getDefaultAccountName(
      networkId,
      type,
    )
  }
  public async getAccount(
    selector: BaseWalletAccount,
  ): Promise<WalletAccount | null> {
    return this.walletAccountSharedService.getAccount(selector)
  }
  public async getSelectedAccount(): Promise<WalletAccount | undefined> {
    return this.walletAccountSharedService.getSelectedAccount()
  }
  public async selectAccount(
    accountIdentifier?: BaseWalletAccount | NetworkOnlyPlaceholderAccount,
  ) {
    return this.walletAccountSharedService.selectAccount(accountIdentifier)
  }
  public async getMultisigAccount(selector: BaseWalletAccount) {
    return this.walletAccountSharedService.getMultisigAccount(selector)
  }
  public async getLastUsedOnNetwork(networkId: string) {
    return this.walletAccountSharedService.getLastUsedAccountOnNetwork(
      networkId,
    )
  }

  // WalletAccountStarknetService
  public async getStarknetAccount(
    selector: BaseWalletAccount,
    useLatest = false,
  ) {
    return this.walletAccountStarknetService.getStarknetAccount(
      selector,
      useLatest,
    )
  }
  public async getSelectedStarknetAccount() {
    return this.walletAccountStarknetService.getSelectedStarknetAccount()
  }
  public async newPendingMultisig(
    networkId: string,
    signerType: SignerType,
  ): Promise<PendingMultisig> {
    return this.walletAccountStarknetService.newPendingMultisig(
      networkId,
      signerType,
    )
  }
  public getStarknetAccountOfType(
    account: Account,
    signer: BaseSignerInterface,
    walletAccount: Pick<WalletAccount, "type" | "guardian">,
  ) {
    return this.walletAccountStarknetService.getStarknetAccountOfType(
      account,
      signer,
      walletAccount,
    )
  }

  public getLedgerAccounts(networkId: string, start: number, total: number) {
    return this.walletAccountStarknetService.getLedgerAccounts(
      networkId,
      start,
      total,
    )
  }

  public addLedgerAccounts(
    accounts: ImportedLedgerAccount[],
    networkId: string,
  ) {
    return this.walletAccountStarknetService.addLedgerAccounts(
      accounts,
      networkId,
    )
  }

  // WalletBackupService
  public async getBackup() {
    return this.walletBackupService.getBackup()
  }
  public async isInitialized() {
    return this.walletBackupService.isInitialized()
  }
  public async importBackup(backup: string) {
    return this.walletBackupService.importBackup(backup)
  }

  // WalletCryptoSharedService
  public async restoreSeedPhrase(seedPhrase: string, newPassword: string) {
    return this.walletCryptoSharedService.restoreSeedPhrase(
      seedPhrase,
      newPassword,
    )
  }
  public async getSeedPhrase(): Promise<string> {
    return this.walletCryptoSharedService.getSeedPhrase()
  }

  // WalletCryptoStarknetService
  public async getArgentPubKeyByDerivationPath(derivationPath: string) {
    return this.walletCryptoStarknetService.getArgentPubKeyByDerivationPath(
      derivationPath,
    )
  }
  public async getSignerForAccount(account: WalletAccount) {
    return this.walletCryptoStarknetService.getSignerForAccount(account)
  }
  public async getPrivateKey(
    baseWalletAccount: BaseWalletAccount,
  ): Promise<string> {
    return this.walletCryptoStarknetService.getPrivateKey(baseWalletAccount)
  }
  public async getPublicKey(baseAccount?: BaseWalletAccount) {
    return this.walletCryptoStarknetService.getPublicKey(baseAccount)
  }

  public async getNextPublicKey(
    accountType: CreateAccountType,
    signerType: SignerType,
    networkId: string,
  ) {
    return this.walletCryptoStarknetService.getNextPublicKey(
      accountType,
      signerType,
      networkId,
    )
  }
  public async getPublicKeysBufferForMultisig(start: number, buffer: number) {
    return this.walletCryptoStarknetService.getPublicKeysBufferForMultisig(
      start,
      buffer,
    )
  }
  public async getUndeployedAccountCairoVersion(
    baseAccount: BaseWalletAccount,
  ) {
    return this.walletCryptoStarknetService.getUndeployedAccountCairoVersion(
      baseAccount,
    )
  }

  public async getCalculatedMultisigAddress(
    baseMultisigAccount: BaseMultisigWalletAccount,
  ) {
    return this.walletCryptoStarknetService.getCalculatedMultisigAddress(
      baseMultisigAccount,
    )
  }
  async getAccountClassHashForNetwork(
    network: Network,
    accountType: ArgentAccountType,
  ): Promise<string> {
    return this.walletCryptoStarknetService.getAccountClassHashForNetwork(
      network,
      accountType,
    )
  }

  // WalletDeploymentStarknetService
  public async deployAccount(
    walletAccount: WalletAccount,
    transactionDetails?: InvocationsDetails | undefined,
  ) {
    return this.walletDeploymentStarknetService.deployAccount(
      walletAccount,
      transactionDetails,
    )
  }
  public async getAccountDeploymentFee(
    walletAccount: WalletAccount,
    feeTokenAddress: Address,
  ) {
    return this.walletDeploymentStarknetService.getAccountDeploymentFee(
      walletAccount,
      feeTokenAddress,
    )
  }
  public async redeployAccount(account: WalletAccount) {
    return this.walletDeploymentStarknetService.redeployAccount(account)
  }
  public async getAccountDeploymentPayload(walletAccount: WalletAccount) {
    return this.walletDeploymentStarknetService.getAccountDeploymentPayload(
      walletAccount,
    )
  }
  public async getMultisigDeploymentPayload(walletAccount: WalletAccount) {
    return this.walletDeploymentStarknetService.getMultisigDeploymentPayload(
      walletAccount,
    )
  }
  public async getAccountOrMultisigDeploymentPayload(
    walletAccount: WalletAccount,
  ) {
    return this.walletDeploymentStarknetService.getAccountOrMultisigDeploymentPayload(
      walletAccount,
    )
  }
  public async getDeployContractPayloadForAccountIndex(
    index: number,
    signerType: SignerType,
    networkId: string,
  ) {
    return this.walletDeploymentStarknetService.getDeployContractPayloadForAccountIndex(
      index,
      signerType,
      networkId,
    )
  }
  public async getDeployContractPayloadForMultisig(props: {
    threshold: number
    signers: string[]
    index: number
    networkId: string
    signerType: SignerType
  }) {
    return this.walletDeploymentStarknetService.getDeployContractPayloadForMultisig(
      props,
    )
  }
  public async newAccount(
    networkId: string,
    type: CreateAccountType = "standard",
    signerType?: SignerType,
    multisigPayload?: MultisigData,
  ) {
    return this.walletDeploymentStarknetService.newAccount(
      networkId,
      type,
      signerType,
      multisigPayload,
    )
  }

  // WalletRecoverySharedService
  public async discoverAccounts() {
    return this.walletRecoverySharedService.discoverAccounts()
  }

  public async restoreMultisigAccountsFromLedger(networkId: string) {
    return this.walletRecoverySharedService.restoreMultisigAccountsFromLedger(
      networkId,
    )
  }

  // WalletSessionService
  public async isSessionOpen() {
    return this.walletSessionService.isSessionOpen()
  }
  public async startSession(
    password: string,
    progressCallback?: ProgressCallback,
  ) {
    return this.walletSessionService.startSession(password, progressCallback)
  }
  public async checkPassword(password: string): Promise<boolean> {
    return this.walletSessionService.checkPassword(password)
  }
  public async lock() {
    return this.walletSessionService.lock()
  }
  public async setSession(secret: string, password: string) {
    return this.walletSessionService.setSession(secret, password)
  }
}
