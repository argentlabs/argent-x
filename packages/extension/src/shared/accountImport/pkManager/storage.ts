import { KeyValueStorage } from "../../storage"
import { adaptKeyValue } from "../../storage/__new/keyvalue"
import type { IPKStore } from "../types"

const keyValueStorage = new KeyValueStorage<IPKStore>(
  {
    keystore: {},
  },
  {
    namespace: "service:pkManager",
  },
)

/**
 * @internal
 * This storage is intended to be used only by the PKManager for storing encrypted keys.
 */
export const pkStore = adaptKeyValue(keyValueStorage)
