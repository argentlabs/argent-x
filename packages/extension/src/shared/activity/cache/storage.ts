import { KeyValueStorage } from "../../storage"
import { adaptKeyValue } from "../../storage/__new/keyvalue"
import type { IActivityCacheStorage } from "./IActivityCacheStorage"

const keyValueStorage = new KeyValueStorage<IActivityCacheStorage>(
  {
    cache: {},
  },
  {
    namespace: "service:activityCache",
  },
)

export const activityCacheStore = adaptKeyValue(keyValueStorage)
