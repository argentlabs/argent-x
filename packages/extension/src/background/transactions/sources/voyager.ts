import join from "url-join"

import { Network } from "../../../shared/network"
import { Transaction, compareTransactions } from "../../../shared/transactions"
import { WalletAccount } from "../../../shared/wallet.model"
import { fetchWithTimeout } from "../../utils/fetchWithTimeout"
import { mapVoyagerTransactionToTransaction } from "../transformers"

export interface VoyagerTransaction {
  blockId: string
  entry_point_type: string | null
  globalIndex?: number
  hash: string
  index: number
  signature: string[] | null
  timestamp: number
  to: string
  type: string
}

export const fetchVoyagerTransactions = async (
  address: string,
  network: Network,
): Promise<VoyagerTransaction[]> => {
  const { explorerUrl } = network
  if (!explorerUrl) {
    return []
  }
  const response = await fetchWithTimeout(
    join(explorerUrl, `api/txns?to=${address}`),
  )
  const { items } = await response.json()
  return items
}

export async function getTransactionHistory(
  accountsToPopulate: WalletAccount[],
  metadataTransactions: Transaction[],
) {
  const accountsWithHistory = accountsToPopulate.filter((account) =>
    Boolean(account.network?.explorerUrl),
  )
  const transactionsPerAccount = await Promise.all(
    accountsWithHistory.map(async (account) => {
      const voyagerTransactions = await fetchVoyagerTransactions(
        account.address,
        account.network,
      )
      return voyagerTransactions.map((transaction) =>
        mapVoyagerTransactionToTransaction(
          transaction,
          account,
          metadataTransactions.find((tx) =>
            compareTransactions(tx, {
              hash: transaction.hash,
              account: { networkId: account.networkId },
            }),
          )?.meta,
        ),
      )
    }),
  )
  return transactionsPerAccount.flat()
}
