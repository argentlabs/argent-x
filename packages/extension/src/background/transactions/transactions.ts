import { uniqWith } from "lodash-es"
import { Status } from "starknet"

import {
  Transaction,
  TransactionBase,
  TransactionRequest,
  compareTransactions,
} from "../../shared/transactions"
import { WalletAccount } from "../../shared/wallet.model"
import { accountsEqual } from "../../shared/wallet.service"
import { getTransactionsStatusUpdate } from "./determineUpdates"
import { getTransactionsUpdate } from "./onchain"
import { setIntervalAsync } from "./setIntervalAsync"
import { GetTransactionsStore } from "./store"
import { checkTransactionHash } from "./transactionExecution"
import { FetchTransactions, getTransactionHistory } from "./voyager"

const timestampInSeconds = (): number => Math.floor(Date.now() / 1000)

export interface TransactionTracker {
  load: (accountsToPopulate: WalletAccount[]) => Promise<void>
  add: (transaction: TransactionRequest) => Promise<void>
  addAccount: (
    account: WalletAccount,
    transaction: TransactionRequest,
  ) => Promise<void>
  get: (transaction: TransactionBase) => Promise<Transaction | null>
  getAll: (statusIn?: Status[]) => Promise<Transaction[]>
  stop: () => void
}

export type TransactionUpdateListener = (
  newTransactions: Transaction[],
  allTransactions: Transaction[],
) => void

type GetTransactionsTracker = (
  getTransactionsStore: GetTransactionsStore,
  fetchTransactions: FetchTransactions,
  onUpdate?: TransactionUpdateListener,
  updateInterval?: number,
) => TransactionTracker

export const getTransactionsTracker: GetTransactionsTracker = (
  getTransactionsStore,
  fetchTransactions,
  onUpdate,
  updateInterval = 15e3, // 15 seconds
  checkHistory = 4 * 5, // check history every 20 update cycles ~ 5 minutes when update interval is 15 seconds
) => {
  const accounts: WalletAccount[] = []
  const transactionsStore = getTransactionsStore()

  let updateCounter = 0
  const handleUpdate = async () => {
    const allTransactions = await transactionsStore.getItems()
    const needsHistoryUpdate = updateCounter === 0
    const historyTransactions = needsHistoryUpdate
      ? await getTransactionHistory(
          accounts,
          allTransactions,
          fetchTransactions,
        )
      : []
    const historyUpdates = getTransactionsStatusUpdate(
      allTransactions,
      historyTransactions,
    )

    const pendingTransactions = allTransactions.filter(
      (tx) =>
        historyTransactions.findIndex((tx2) => compareTransactions(tx, tx2)) ===
        -1,
    )

    const onChainUpdates = await getTransactionsUpdate(pendingTransactions)

    const updates = [...onChainUpdates, ...historyUpdates]

    await transactionsStore.addItems(updates)
    onUpdate?.(updates, allTransactions)

    updateCounter = (updateCounter + 1) % checkHistory // as this is done at the very end, onerror it will not be incremented and therefore keep the history cycle refreshing at 15 seconds
  }

  const clearUpdate = setIntervalAsync(handleUpdate, updateInterval)

  const add = async (transaction: TransactionRequest) => {
    // sanity checks
    if (!checkTransactionHash(transaction.hash)) {
      return // dont throw
    }

    const newTransaction = {
      status: "RECEIVED" as const,
      timestamp: timestampInSeconds(),
      ...transaction,
    }

    const allTransactions = await transactionsStore.getItems()
    await transactionsStore.addItem(newTransaction)

    onUpdate?.([newTransaction], allTransactions)
  }

  return {
    load: async (accountsToPopulate) => {
      const allTransactions = await transactionsStore.getItems()
      const initialAccounts = uniqWith(accountsToPopulate, accountsEqual)
      const initialTransactions = await getTransactionHistory(
        initialAccounts,
        allTransactions,
        fetchTransactions,
      )
      transactionsStore.addItems(initialTransactions)
      accounts.splice(0, accounts.length, ...initialAccounts)
    },
    add,
    addAccount: async (account, transaction) => {
      if (
        !accounts.find((existingAccount) =>
          accountsEqual(existingAccount, account),
        )
      ) {
        accounts.push(account)
      }
      await add(transaction)
    },
    get: (transaction: TransactionBase) =>
      transactionsStore.getItem((tx) => compareTransactions(tx, transaction)),
    getAll: (statusIn) =>
      transactionsStore.getItems(
        statusIn
          ? (transaction) => statusIn.includes(transaction.status)
          : undefined,
      ),
    stop: clearUpdate,
  }
}
