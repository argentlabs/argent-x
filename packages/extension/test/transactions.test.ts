import "isomorphic-fetch"

import { Call } from "starknet"
import { afterEach, beforeEach, describe, expect, test, vi } from "vitest"
import waitForExpect from "wait-for-expect"

import { ArrayStorage } from "../src/background/storage/array"
import { getTransactionsStatusUpdate } from "../src/background/transactions/determineUpdates"
import { setIntervalAsync } from "../src/background/transactions/setIntervalAsync"
import { nameTransaction } from "../src/background/transactions/transactionNames"
import {
  TransactionTracker,
  getTransactionsTracker,
} from "../src/background/transactions/transactions"
import { FetchTransactions } from "../src/background/transactions/voyager"
import { defaultNetwork } from "../src/shared/networks"
import {
  Transaction,
  TransactionRequest,
  compareTransactions,
} from "../src/shared/transactions"
import { WalletAccount } from "../src/shared/wallet.model"
import { MockStorage } from "./mock"
import transactionsMock from "./transactions.mock.json"

const wallet: WalletAccount = {
  address: "0x05e54edb59e1b1e398f9647e617276f6da0eb9ddfc0c02723269b9baa2489dce",
  networkId: defaultNetwork.id,
  network: defaultNetwork,
  signer: {
    derivationPath: "0",
    type: "local_secret",
  },
}

const wait = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms))

export const getTransactionsStore = () =>
  new ArrayStorage<Transaction>(
    [],
    new MockStorage({ inner: [] }),
    compareTransactions,
  )

const fetchMockTransactions: FetchTransactions = async (_, __) =>
  transactionsMock.items

describe("transactions", () => {
  let txTracker: TransactionTracker
  const fn = vi.fn()
  beforeEach(async () => {
    txTracker = getTransactionsTracker(
      getTransactionsStore,
      fetchMockTransactions,
      fn,
      1000,
    )
    await txTracker.load([wallet])
  })
  afterEach(() => {
    fn.mockReset()
    txTracker.stop()
  })
  test("should get all transactions when initializing", async () => {
    const transactions = await txTracker.getAll()
    expect(transactions.length).toBeGreaterThan(5)
    expect(fn).toBeCalledTimes(0)
  })
  test("should get a transaction by hash", async () => {
    const transaction = await txTracker.get({
      hash: "0x16d38d961a659b7565b596060ca812b863a39766accab5a8fd93ace56e6001a",
      account: {
        networkId: wallet.networkId,
      },
    })
    expect(transaction).toMatchSnapshot()
    expect(fn).toBeCalledTimes(0)
  })
  test("should add a transaction", async () => {
    const transactionHash =
      "0x16d38d961a659b7565b596060ca812b863a39766accab5a8fd93ace56e6001a" // add a transaction which already exists onchain
    const transaction: TransactionRequest = {
      account: wallet,
      hash: transactionHash,
      meta: {
        title: "test",
        subTitle: "test",
      },
    }
    await txTracker.add(transaction)
    const transactions = await txTracker.get({
      hash: transactionHash,
      account: { networkId: wallet.networkId },
    })
    expect(transactions).toMatchObject({
      // after being added, the transaction should be in the store with the default status
      ...transaction,
      status: "RECEIVED",
      timestamp: expect.any(Number),
    })
    expect(fn).toBeCalledTimes(1)
    expect(fn).toBeCalledWith([
      { ...transaction, status: "RECEIVED", timestamp: expect.any(Number) }, // after the update , the transaction should be accepted on L2
    ])

    await waitForExpect(() => {
      expect(fn).toBeCalledTimes(2)
    }, 2000)
    expect(fn).toBeCalledWith([
      { ...transaction, status: "ACCEPTED_ON_L2", timestamp: 1652257464 }, // after the update , the transaction should be accepted on L2
    ])

    await waitForExpect(() => {
      expect(fn).toBeCalledTimes(3)
    }, 2000)
    expect(fn).toBeCalledWith([]) // no update this time, as ACCEPTED_ON_L2 us considered as final status
  })
})

describe("getTransactionsStatusUpdate()", () => {
  const transaction1: Transaction = {
    account: wallet,
    hash: "0x16d38d961a659b7565b596060ca812b863a39766accab5a8fd93ace56e6001a",
    meta: {
      title: "test",
      subTitle: "test",
    },
    status: "RECEIVED",
    timestamp: 1001,
  }
  const transaction2: Transaction = {
    account: wallet,
    hash: "0x26d38d961a659b7565b596060ca812b863a39766accab5a8fd93ace56e6001b",
    meta: {
      title: "test2",
      subTitle: "test2",
    },
    status: "RECEIVED",
    timestamp: 1001,
  }

  test("compare 2 transactions to 2 updated transactions expect 1 change", async () => {
    const transactions: Transaction[] = [transaction1, transaction2]
    const updatedTransactions: Transaction[] = [
      transaction1,
      {
        ...transaction2,
        status: "PENDING",
      },
    ]

    const change = getTransactionsStatusUpdate(
      transactions,
      updatedTransactions,
    )

    expect(change.length).toBe(1)
  })

  test("compare 2 transactions to 1 updated transactions expect 1 change", async () => {
    const transactions: Transaction[] = [transaction1, transaction2]
    const updatedTransactions: Transaction[] = [
      {
        ...transaction2,
        status: "ACCEPTED_ON_L2",
      },
    ]

    const change = getTransactionsStatusUpdate(
      transactions,
      updatedTransactions,
    )

    expect(change.length).toBe(1)
  })

  test("compare 2 transactions to 0 updated transactions expect 0 change", async () => {
    const transactions: Transaction[] = [transaction1, transaction2]
    const updatedTransactions: Transaction[] = []

    const change = getTransactionsStatusUpdate(
      transactions,
      updatedTransactions,
    )

    expect(change.length).toBe(0)
  })

  test("compare 0 transactions to 2 updated transactions expect 2 change", async () => {
    const transactions: Transaction[] = []
    const updatedTransactions: Transaction[] = [transaction1, transaction2]

    const change = getTransactionsStatusUpdate(
      transactions,
      updatedTransactions,
    )

    expect(change.length).toBe(2)
  })
})

describe("setAsyncInterval()", () => {
  test("should run fn every n seconds", async () => {
    const fn = vi.fn()
    const n = 100

    const stop = setIntervalAsync(fn, n)
    expect(fn).toHaveBeenCalledTimes(0)
    await wait(n)
    expect(fn).toHaveBeenCalledTimes(1)
    await wait(n / 2)
    expect(fn).toHaveBeenCalledTimes(1)
    await wait(n / 2)
    expect(fn).toHaveBeenCalledTimes(2)

    stop()
    await wait(n * 2)
    expect(fn).toHaveBeenCalledTimes(2)
  })
})

describe("nameTransaction()", () => {
  test("should return right metadata for 1 transaction", () => {
    const transactions: Call[] = [
      { entrypoint: "transfer_from", contractAddress: "0x0", calldata: [] },
    ]

    const metadata = nameTransaction(transactions)

    expect(metadata).toMatchSnapshot()
  })
  test("should return right metadata for 2 transactions", () => {
    const transactions: Call[] = [
      { entrypoint: "approve", contractAddress: "0x0", calldata: [] },
      { entrypoint: "transfer", contractAddress: "0x0", calldata: [] },
    ]

    const metadata = nameTransaction(transactions)

    expect(metadata).toMatchSnapshot()
  })
})
