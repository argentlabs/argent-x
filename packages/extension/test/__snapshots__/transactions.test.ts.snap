// Vitest Snapshot v1

exports[`nameTransaction() > should return right metadata for 1 transaction 1`] = `
{
  "title": "Transfer from",
}
`;

exports[`nameTransaction() > should return right metadata for 2 transactions 1`] = `
{
  "title": "Approve and Transfer",
}
`;

exports[`transactions > should get a transaction by hash 1`] = `
{
  "account": {
    "address": "0x05e54edb59e1b1e398f9647e617276f6da0eb9ddfc0c02723269b9baa2489dce",
    "network": {
      "accountClassHash": "0x3e327de1c40540b98d05cbcb13552008e36f0ec8d61d46956d2f9752c294328",
      "baseUrl": "https://alpha4.starknet.io",
      "chainId": "SN_GOERLI",
      "explorerUrl": "https://goerli.voyager.online",
      "id": "goerli-alpha",
      "multicallAddress": "0x042a12c5a641619a6c58e623d5735273cdfb0e13df72c4bacb4e188892034bd6",
      "name": "Goerli Testnet",
      "readonly": true,
    },
    "networkId": "goerli-alpha",
    "signer": {
      "derivationPath": "0",
      "type": SignerType.LOCAL_SECRET,
    },
  },
  "hash": "0x16d38d961a659b7565b596060ca812b863a39766accab5a8fd93ace56e6001a",
  "meta": undefined,
  "status": "ACCEPTED_ON_L2",
  "timestamp": 1652257464,
}
`;
